javalib/src/main/scala/java/util/ArrayDeque.scala

```diff
@@ -171,55 +193,175 @@ class ArrayDeque[E] private (private val inner: js.Array[E])
+        if (lastIndex == -1) {
           throw new IllegalStateException()
         } else {
-          arrayRemove(inner, index)
+          val laterShifted = removeAt(lastIndex)
+          expectedStatus = self.status
+          if (forward && laterShifted) {
+            nextIndex -= 1
+          }
+
+          if (!forward && !laterShifted) {
+            nextIndex += 1
+          }
         }
       }
     }
   }
 
-  def iterator(): Iterator[E] =
-    failFastIterator(-1, x => (x + 1))
+  def iterator(): Iterator[E] = failFastIterator(true)
 
-  def descendingIterator(): Iterator[E] =
-    failFastIterator(inner.length, x => (x - 1))
+  def descendingIterator(): Iterator[E] = failFastIterator(false)
 
-  override def contains(o: Any): Boolean =
-    arrayExists(inner)(Objects.equals(_, o))
+  override def contains(o: Any): Boolean = firstIndexOf(o) != -1
 
   override def remove(o: Any): Boolean = removeFirstOccurrence(o)
 
   override def clear(): Unit = {
     if (!isEmpty())
       status += 1
-    inner.length = 0
+    empty = true
+    startIndex = 0
+    endIndex = inner.length
+  }
+
+  private def firstIndexOf(o: Any): Int = {
+    // scalastyle:off return
+    if (isEmpty())
+      return -1
+    val inner = this.inner // local copy
+    val capacity = inner.length // local copy
+    val endIndex = this.endIndex // local copy
+    var i = startIndex
+    do {
+      if (i >= capacity)
+        i = 0
+      if (Objects.equals(inner(i), o)) {
+        return i
+      }
+      i += 1 // let i overrun so we catch endIndex == capacity
+    } while (i != endIndex)
+    -1
+    // scalastyle:on return
+  }
+
+  private def ensureCapacityForAdd(): Unit = {
+    if (isEmpty()) {
+      // Nothing to do (constructor ensures capacity is always non-zero).
+    } else if (startIndex == 0 && endIndex == inner.length) {
+      inner.length *= 2
+      // no copying required: We just keep adding to the end.
+    } else if (startIndex == endIndex) {
+      val oldCapacity = inner.length
+      inner.length *= 2
+      // move beginning of array to end
+      var i = 0
+      while (i != endIndex) {
+        inner(i + oldCapacity) = inner(i)
+        i += 1
+      }
+      endIndex += oldCapacity
+    }
+  }
+
+  /* Removes the element at index [[target]]
+   *
+   * @returns true if elements after the target index were shifted.
+   * Returns false, if elements before the target index were shifted.
+   */
+  private def removeAt(target: Int): Boolean = {
+    if (target == startIndex) {
+      pollFirst()
+      false
+    } else if (target == endIndex - 1) {
+      pollLast()
+      true
+    } else if (startIndex < endIndex || target < endIndex) {
+      // Shift elements from endIndex towards target
+      var i = target
+      while (i != endIndex - 1) {
+        inner(i) = inner(i + 1)
+        i += 1
+      }
+      status += 1
+      endIndex -= 1
+      if (startIndex == endIndex)
+        empty = true
```

Comment:
If this happens, it means that there was exactly 1 element in the queue before `removeAt`. But then, it must have been the case that `target == startIndex` when entering the method, and therefore the `if (target == startIndex)` branch at the beginning would have been taken. So this will never happen.